cmd/clusterctl/internal/migrate/parser.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migrate
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	jsonserializer "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
+)
+
+// Document represents a single YAML document with associated metadata.
+type Document struct {
+	Object runtime.Object
+	Raw    []byte
+	GVK    schema.GroupVersionKind
+	Type   ResourceType
+	// Index indicates the document's position in the original stream.
+	Index int
+}
+
+// ResourceType classifies the type of Kubernetes resource.
+type ResourceType int
+
+const (
+	// ResourceTypeCoreV1Beta1 identifies v1beta1 core ClusterAPI resources.
+	ResourceTypeCoreV1Beta1 ResourceType = iota
+	// ResourceTypeOtherCAPI identifies other cluster.x-k8s.io resources (non-v1beta1).
+	ResourceTypeOtherCAPI
```

Comment:
what about
```suggestion
	ResourceTypeCoreOtherVersion
```

Reply:
I realize we can generalise this. Refactoring is incoming. Thanks